File: raffia-core/src/main/java/com/codepoetics/raffia/operations/Updaters.java
package com.codepoetics.raffia.operations;

import com.codepoetics.raffia.baskets.ArrayContents;
import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.baskets.PropertySet;
import com.codepoetics.raffia.mappers.Mapper;

import java.math.BigDecimal;

public final class Updaters {

  public static final Updater NO_OP = new Updater() {
    @Override
    public Basket update(Basket basket) {
      return basket;
    }
  };

  public static Updater branch(final BasketPredicate predicate, final Updater ifTrue, final Updater ifFalse) {
    return new Updater() {
      @Override
      public Basket update(Basket basket) {
        return predicate.test(basket)
            ? ifTrue.update(basket)
            : ifFalse.update(basket);
      }
    };
  }

  public static Updater toConstant(final Basket value) {
    return new Updater() {
      @Override
      public Basket update(Basket basket) {
        return value;
      }
    };
  }

  public static Updater ofString(final Mapper<String, String> stringMapper) {
    return new Updater() {
      @Override
      public Basket update(Basket basket) {
        return basket.isString() ? Basket.ofString(stringMapper.map(basket.asString())) : basket;
      }
    };
  }

  public static Updater ofNumber(final Mapper<BigDecimal, BigDecimal> numberMapper) {
    return new Updater() {
      @Override
      public Basket update(Basket basket) {
        return basket.isNumber() ? Basket.ofNumber(numberMapper.map(basket.asNumber())) : basket;
      }
    };
  }

  public static Updater ofBoolean(final Mapper<Boolean, Boolean> booleanMapper) {
    return new Updater() {
      @Override
      public Basket update(Basket basket) {
        return basket.isBoolean() ? Basket.ofBoolean(booleanMapper.map(basket.asBoolean())) : basket;
      }
    };
  }

  public static Updater ofArray(final Mapper<ArrayContents, ArrayContents> arrayContentsMapper) {
    return new Updater() {
      @Override
      public Basket update(Basket basket) {
        return basket.mapArray(arrayContentsMapper);
      }
    };
  }

  public static Updater ofObject(final Mapper<PropertySet, PropertySet> propertySetMapper) {
    return new Updater() {
      @Override
      public Basket update(Basket basket) {
        return basket.mapObject(propertySetMapper);
      }
    };
  }

  public static Updater appending(final Basket arrayItem) {
    return ofArray(new Mapper<ArrayContents, ArrayContents>() {
      @Override
      public ArrayContents map(ArrayContents input) {
        return input.plus(arrayItem);
      }
    });
  }

  public static Updater inserting(final int index, final Basket arrayItem) {
    return ofArray(new Mapper<ArrayContents, ArrayContents>() {
      @Override
      public ArrayContents map(ArrayContents input) {
        return input.plus(index, arrayItem);
      }
    });
  }

  public static Updater replacing(final int index, final Basket arrayItem) {
    return ofArray(new Mapper<ArrayContents, ArrayContents>() {
      @Override
      public ArrayContents map(ArrayContents input) {
        return input.with(index, arrayItem);
      }
    });
  }

  public static Updater inserting(final String key, final Basket value) {
    return ofObject(new Mapper<PropertySet, PropertySet>() {
      @Override
      public PropertySet map(PropertySet input) {
        return input.with(key, value);
      }
    });
  }

  public static Updater removing(final int index) {
    return ofArray(new Mapper<ArrayContents, ArrayContents>() {
      @Override
      public ArrayContents map(ArrayContents input) {
        return input.minus(index);
      }
    });
  }

  public static Updater removing(final String key) {
    return ofObject(new Mapper<PropertySet, PropertySet>() {
      @Override
      public PropertySet map(PropertySet input) {
        return input.minus(key);
      }
    });
  }

  public static Updater updating(final int index, final Updater itemUpdater) {
    return ofArray(new Mapper<ArrayContents, ArrayContents>() {
      @Override
      public ArrayContents map(ArrayContents input) {
        int actual = index < 0 ? input.size() + index : index;
        if (actual < 0 || actual >= input.size()) {
          return input;
        }
        return input.with(actual, itemUpdater.update(input.get(actual)));
      }
    });
  }

  public static Updater updating(final String key, final Updater valueUpdater) {
    return ofObject(new Mapper<PropertySet, PropertySet>() {
      @Override
      public PropertySet map(PropertySet input) {
        return input.containsKey(key) ? input.with(key, valueUpdater.update(input.get(key))) : input;
      }
    });
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/operations/Updaters.kt
package com.codepoetics.raffia.operations

import com.codepoetics.raffia.baskets.ArrayContents
import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.baskets.PropertySet

import java.math.BigDecimal

typealias BasketPredicate = (Basket) -> Boolean
typealias Updater = (Basket) -> Basket

object Updaters {

    val NO_OP: Updater = { basket -> basket }

    @JvmStatic
    fun branch(predicate: BasketPredicate, ifTrue: Updater, ifFalse: Updater): Updater = { basket ->
        if (predicate(basket)) ifTrue(basket)
        else ifFalse(basket)
    }

    @JvmStatic
    fun toConstant(value: Basket): Updater = { value }

    @JvmStatic
    fun ofString(stringMapper: (String) -> String): Updater = { basket ->
        if (basket.isString()) basket.mapString(stringMapper) else basket
    }

    @JvmStatic
    fun ofNumber(numberMapper: (BigDecimal) -> BigDecimal): Updater = { basket ->
        if (basket.isNumber()) basket.mapNumber(numberMapper) else basket
    }

    @JvmStatic
    fun ofBoolean(booleanMapper: (Boolean) -> Boolean): Updater = { basket ->
        if (basket.isBoolean()) basket.mapBoolean(booleanMapper) else basket
    }

    @JvmStatic
    fun ofArray(arrayContentsMapper: (ArrayContents) -> ArrayContents): Updater = { basket ->
        if (basket.isArray()) basket.mapArray(arrayContentsMapper) else basket
    }

    @JvmStatic
    fun ofObject(propertySetMapper:(PropertySet) -> PropertySet): Updater = { basket ->
        if (basket.isObject()) basket.mapObject(propertySetMapper) else basket
    }

    @JvmStatic
    fun appending(arrayItem: Basket): Updater = ofArray { input -> input.plus(arrayItem) }

    @JvmStatic
    fun inserting(index: Int, arrayItem: Basket): Updater = ofArray { input -> input.plus(index, arrayItem) }

    @JvmStatic
    fun replacing(index: Int, arrayItem: Basket): Updater = ofArray { input -> input.with(index, arrayItem) }

    @JvmStatic
    fun inserting(key: String, value: Basket): Updater = ofObject { input -> input.with(key, value) }

    @JvmStatic
    fun removing(index: Int): Updater = ofArray { input -> input.minus(index) }

    @JvmStatic
    fun removing(key: String): Updater = ofObject { input -> input.minus(key) }

    @JvmStatic
    fun updating(index: Int, itemUpdater: Updater): Updater = ofArray { input ->
        val actual = if (index < 0) input.size() + index else index
        if (actual < 0 || actual >= input.size()) {
            return@ofArray input
        }
        input.with(actual, itemUpdater(input[actual]))
    }

    @JvmStatic
    fun updating(key: String, valueUpdater: Updater): Updater = ofObject { input ->
        if (input.containsKey(key)) input.with(key, valueUpdater(input[key]!!))
        else input
    }
}
